### DISCUSSION TITLE
Casting from a `TypedDict` to a `Mapping` can be a bit more lax

### DISCUSSION THREAD

**Discussion by tusharsadhwani** [microsoft/pyright#9513] [Ideas]
Currently, this code fails type checking:

```python
from collections.abc import Mapping
from typing import TypedDict

def my_func(v: Mapping[str, str]):
    print(v.items())

class MyTypedDict(TypedDict):
    foo: str

val: MyTypedDict = {"foo": "a"}
my_func(val)
```

Changing the annotation of `v` to `Mapping[str, object]` works. But since the TypedDict only contains `str` fields, I think this cast should work? Is there a reason for the cast to not work right now?

**Comment by erictraut** [comment]
The typing spec specifically disallows assignment to a more specific `Mapping` type because this is not type safe. You can read about this in the official typing spec [here](https://typing.readthedocs.io/en/latest/spec/typeddict.html#assignability). A `TypedDict` is an "open" structural type, and a compatible dictionary is allowed to contain keys that are not specified in the `TypedDict` definition. 

There is a [draft PEP](https://peps.python.org/pep-0728/) that defines support for "closed" TypedDicts. This PEP hasn't yet been approved, but pyright contains provisional support for the latest draft. You can try it by setting `enableExperimentalFeatures` to true in your pyright configuration.
